In [33]:
from sklearn.datasets import load_iris
from sklearn.linear_model import Perceptron
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.neural_network import MLPClassifier



In [2]:
from sklearn.linear_model import Perceptron


iris = load_iris(as_frame=True) # or False if you prefer NumPy arrays

In [3]:
X = iris.data
y = iris.target

In [4]:
# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42)


In [5]:
# Initializing the Perceptron model
perceptron = Perceptron()

In [6]:
perceptron.fit(X_train, y_train)

Perceptron()

In [7]:
# model prediction
y_pred = perceptron.predict(X_test)


In [8]:
# Evaluating the model performance
accuracy = accuracy_score(y_test, y_pred)
print("Perceptron accuracy:", accuracy)

Perceptron accuracy: 0.8222222222222222


In [23]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
y_test_encoded = le.fit_transform(y_test)
y_pred_encoded = le.transform(y_pred)


In [10]:
correct = [0] * len(set(y_test_encoded))
for i in range(len(y_test_encoded)):
    if y_test_encoded[i] == y_pred_encoded[i]:
        correct[y_test_encoded[i]] += 1
print("Correct predictions per class:", correct)

Correct predictions per class: [19, 5, 13]


In [11]:
X_new = [[2, 0.5, 1, 1], [3, 1, 0, 0.5]]
y_pred = perceptron.predict(X_new)
y_pred

/usr/local/lib/python3.9/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but Perceptron was fitted with feature names
  warnings.warn(


array([1, 1])

In [13]:
# Defining the  grid
param_grid = {
    "max_iter": [10, 50, 100, 500, 1000]}

In [14]:
param_grid

{'max_iter': [10, 50, 100, 500, 1000]}

In [17]:
# Initializing the grid search
grid_search = GridSearchCV(
    estimator=Perceptron(), param_grid=param_grid, cv=5)

In [18]:
grid_search.fit(X_train, y_train)

/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_stochastic_gradient.py:702: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn(
/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_stochastic_gradient.py:702: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn(
/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_stochastic_gradient.py:702: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn(
/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_stochastic_gradient.py:702: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn(
/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_stochastic_grad

GridSearchCV(cv=5, estimator=Perceptron(),
             param_grid={'max_iter': [10, 50, 100, 500, 1000]})

In [19]:
# Printing the best hyperparameters
print("Best hyperparameters:", grid_search.best_params_)

Best hyperparameters: {'max_iter': 10}


In [29]:
y_pred_grid = grid_search.predict(X_test)

In [27]:
# Evaluate the model performance
accuracy = accuracy_score(y_test, y_pred_grid)
print("Perceptron accuracy:", accuracy)

Perceptron accuracy: 0.7111111111111111


In [30]:
#label encoding
y_pred_encoded_grid = le.transform(y_pred_grid)

In [31]:
# Count the number of correct predictions per class
correct = [0] * len(set(y_test_encoded))
for i in range(len(y_test_encoded)):
    if y_test_encoded[i] == y_pred_encoded_grid[i]:
        correct[y_test_encoded[i]] += 1
print("Correct predictions per class:", correct)

Correct predictions per class: [19, 0, 13]


In [34]:
# Create the MLP classifier model
model = MLPClassifier(hidden_layer_sizes=(10,), max_iter=1000, random_state=42)

In [35]:
# Fit the model to the training data
model.fit(X_train, y_train)

/usr/local/lib/python3.9/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


MLPClassifier(hidden_layer_sizes=(10,), max_iter=1000, random_state=42)

In [36]:
# Evaluate the model's performance on the testing data
score = model.score(X_test, y_test)
print(f"MLP classifier accuracy: {score:.2f}")

MLP classifier accuracy: 0.98


Conclusion:

The MLP classifier has a higher accuracy of 0.98 compared to the Perceptron's accuracy of 0.82 on the same dataset. This is because the MLP classifier is a more complex and powerful model compared to the Perceptron. The MLP classifier has hidden layers that allow it to learn non-linear relationships in the data, while the Perceptron only has a single layer and can only learn linear decision boundaries. Therefore, the MLP classifier is able to capture more complex patterns in the data, resulting in higher accuracy.